src/drivers/pca9685_pwm/PCA9685.cpp

```diff
@@ -0,0 +1,292 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2017 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include <cinttypes>
+#include <px4_log.h>
+#include <cmath>
+#include "PCA9685.h"
+
+using namespace drv_pca9685_pwm;
+
+PCA9685::PCA9685(int bus, int addr):
+	I2C("PCA9685", PWM_OUTPUT0_DEVICE_PATH, bus, addr, 400000)
+{
+
+}
+
+int PCA9685::Start()
+{
+	int ret = init();
+	return ret;
+}
+
+int PCA9685::Stop()
+{
+	disAll();
+	return PX4_OK;
+}
+
+int PCA9685::updatePWM(const uint16_t *outputs, unsigned num_outputs)
+{
+	if (num_outputs > PCA9685_PWM_CHANNEL_COUNT) {
+		num_outputs = PCA9685_PWM_CHANNEL_COUNT;
+		PX4_WARN("PCA9685 can only drive up to 16 channels");
+	}
+
+	uint16_t out[PCA9685_PWM_CHANNEL_COUNT];
+	memcpy(out, outputs, sizeof(uint16_t) * num_outputs);
+
+	for (unsigned i = 0; i < num_outputs; ++i) {
+		out[i] = (uint16_t)roundl((out[i] * _Freq * (double)PCA9685_PWM_RES / 1e6)); // convert us to 12 bit resolution
+	}
+
+	setPWM(num_outputs, out);
+
+	return 0;
+}
+
+int PCA9685::setFreq(int freq)
+{
+	uint16_t _output;
+	_output = floorl((double)PCA9685_CLOCK_FREQ / freq);
```

Comment:
You can put that on one line:

```suggestion
	uint16_t _output = floorl((double)PCA9685_CLOCK_FREQ / freq);
```

Also, the leading underscore is usually used for class variables and not local variables.